src/std_lock_api.rs

```diff
@@ -0,0 +1,225 @@
+//! Implements the [`StdRawRwLock`] helper type which wraps a
+//! [`std::syn::RwLock`] so that it can be used with `lock_api`
+
+use {
+    core::ops::Deref,
+    std::{
+        cell::UnsafeCell,
+        fmt,
+        mem::MaybeUninit,
+        pin::Pin,
+        sync::{Once, RwLock, RwLockWriteGuard, TryLockError},
+    },
+};
+
+use lock_api::RawRwLock;
+
+pub struct StdRawRwLock {
+    lock: LazyPinBox<RwLock<()>>,
+    write: UnsafeCell<Option<RwLockWriteGuard<'static, ()>>>,
+}
+
+#[allow(unsafe_code)]
+// Safety:
+// - RwLockWriteGuard is !Send
+// - write is only accessed through RawRwLock while an exclusive lock is held
+// - RawRwLock::GuardMarker is set to GuardNoSend
+// - StdRawRwLock can be Send since the guard exposed through RawRwLock is !Send
+unsafe impl Send for StdRawRwLock {}
+
+#[allow(unsafe_code)]
+// Safety:
+// - UnsafeCell is !Sync
+// - write is only accessed through RawRwLock while an exclusive lock is held
+// - thus StdRawRwLock can be sync
+unsafe impl Sync for StdRawRwLock {}
+
+impl fmt::Debug for StdRawRwLock {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.debug_struct("StdRawRwLock").finish()
+    }
+}
+
+impl Drop for StdRawRwLock {
+    fn drop(&mut self) {
+        // Drop any outstanding write guard before the lock
+        std::mem::drop(self.write.get_mut().take());
+    }
+}
+
+#[allow(unsafe_code)]
+// Safety: std::sync::RwLock fulfills the safety contract of RawRwLock
+unsafe impl RawRwLock for StdRawRwLock {
+    #[allow(clippy::declare_interior_mutable_const)]
+    const INIT: Self = StdRawRwLock {
+        lock: LazyPinBox::new(),
+        write: UnsafeCell::new(None),
+    };
+
+    type GuardMarker = lock_api::GuardNoSend;
+
+    fn lock_shared(&self) {
+        let guard = self
+            .lock
+            .read()
+            .expect("interner lock should not be poisoned");
+
+        // The read guard can be reconstructed later from another read
+        std::mem::forget(guard);
+    }
+
+    fn try_lock_shared(&self) -> bool {
+        let guard = match self.lock.try_read() {
+            Ok(guard) => guard,
+            Err(TryLockError::WouldBlock) => return false,
+            r @ Err(TryLockError::Poisoned(_)) => r.expect("interner lock should not be poisoned"),
+        };
+
+        // The read guard can be reconstructed later from another read
+        std::mem::forget(guard);
+
+        true
+    }
+
+    unsafe fn unlock_shared(&self) {
+        // Since this method may only be called if a shared lock is held,
+        // this if branch should always be taken
+        if let Ok(new_guard) = self.lock.try_read() {
+            // Safety:
+            // - unlock_shared may only be called if a shared lock is held
+            //   in the current context
+            // - thus an old guard for that shared lock must exist
+            let old_guard = std::ptr::read(&new_guard);
```

Comment:
I'm very unconvinced that this is guaranteed to work as intended. I agree that it *does* work without UB *for the current std distribution*, but this isn't guaranteed to be the case forever, e.g. if `RwLockReadGuard` owns some allocation in the future rather than just borrowing the `RwLock`.

I'd be more comfortable if `lock_api` provided a raw version of the std lock for use with `lock_api`. That they don't optionally provide such suggests that they agree with me that this isn't sound.